src/BCFKeyValueData.cc

```diff
@@ -156,142 +157,23 @@ Status BCFKeyValueData::sample_dataset(const string& sample, string& ans) const
-            if (!hdr) {
-                return Status::IOError("BCFReader::Open bcf_hdr_read");
-            }
-        }
-        ans->hdr_ = hdr;
-
-        return Status::OK();
-    }
-
-    virtual ~BCFReader() {
-        if (bcf_) {
-            bcf_close(bcf_);
-        }
-    }
-
-    shared_ptr<const bcf_hdr_t> header() const {
-        assert(hdr_);
-        return hdr_;
-    }
-
-    Status read(shared_ptr<bcf1_t>& ans) {
-        if (!ans) {
-            ans = shared_ptr<bcf1_t>(bcf_init(), &bcf_destroy);
-        }
-        // TODO use bcf_read; need to override htsFile format detection
-        switch (vcf_read(bcf_, hdr_.get(), ans.get())) {
-            case 0: return Status::OK();
-            case -1: return Status::NotFound();
-        }
-        return Status::IOError("BCFReader::read bcf_read");
-    }
-};
-
-class BCFWriter {
-    bcf_hdr_t *hdr_;
-    vcfFile *bcf_ = nullptr;
-    char *buf_ = nullptr;
-    size_t bufsz_;
-
-    BCFWriter() = default;
-
-public:
-
-    static Status Open(bcf_hdr_t *hdr, bool write_header, unique_ptr<BCFWriter>& ans) {
-        ans.reset(new BCFWriter);
-        ans->hdr_ = hdr;
-
-        char fn[4+sizeof(void**)+sizeof(size_t*)];
-        char *pfn = fn;
-
-        memcpy(pfn, "mem:", 4);
-        *(char***)(pfn+4) = &(ans->buf_);
-        *(size_t**)(pfn+4+sizeof(void**)) = &(ans->bufsz_);
-
-        ans->bcf_ = bcf_open(pfn, "wu");
-        if (!ans->bcf_) return Status::Failure("BCFWriter::Open");
-
-        if (write_header) {
-            if (bcf_hdr_write(ans->bcf_, hdr) != 0) {
-                return Status::IOError("BCFWriter::Open bcf_hdr_write");
-            }
-        }
-
-        return Status::OK();
-    }
-
-    virtual ~BCFWriter() {
-        if (bcf_) {
-            bcf_close(bcf_);
-        }
-        if (buf_) {
-            free(buf_);
-        }
-    }
-
-    Status write(bcf1_t* x) {
-        if (bcf_write(bcf_, hdr_, x) != 0) {
-            return Status::IOError("BCFWriter::write");
-        }
-        return Status::OK();
-    }
-
-    Status contents(string& ans) {
-        if (hflush(bcf_->fp.hfile) != 0) {
-            return Status::IOError("BCFWriter::buffer hflush");
-        }
-        size_t sz = htell(bcf_->fp.hfile);
-        ans.clear();
-        if (buf_) {
-            ans = string(buf_, sz);
-        }
-        return Status::OK();
-    }
-};
-
 Status BCFKeyValueData::dataset_bcf_header(const string& dataset,
                                            shared_ptr<const bcf_hdr_t>& hdr) const {
-    // Retrieve the header
-    Status s;
-    KeyValue::CollectionHandle coll;
-    S(body_->db->collection("header",coll));
-    string data;
-    S(body_->db->get(coll, dataset, data));
-
-    // Parse the header
-    unique_ptr<BCFReader> reader;
-    S(BCFReader::Open(nullptr, data.c_str(), data.size(), reader));
-    hdr = reader->header();
-    assert(hdr);
+    if (!hdr) {
```

Comment:
True. 

This was done to accommodate the programming style of calling this with a header already read beforehand.